### ISSUE TITLE
Poco unable to send content and file attachment simultaneously in email

### ISSUE THREAD

**Issue by Cupidvogel** [pocoproject/poco#815]
I am trying to send an email using Poco. I am using my Gmail auth details to configure the sender. Here's the code:

```C++
#include <iostream>

#include "Poco/Net/MailMessage.h"
#include "Poco/Net/MailRecipient.h"
#include "Poco/Net/SMTPClientSession.h"
#include "Poco/Net/NetException.h"
#include "Poco/Net/SecureSMTPClientSession.h"
#include "Poco/Net/InvalidCertificateHandler.h"
#include "Poco/Net/AcceptCertificateHandler.h"
#include "Poco/Net/SSLManager.h"
#include "Poco/Net/SecureStreamSocket.h"
#include "Poco/Net/MailRecipient.h"

using Poco::Net::InvalidCertificateHandler;
using Poco::Net::AcceptCertificateHandler;
using Poco::Net::Context;
using Poco::Net::SSLManager;
using Poco::Net::SecureStreamSocket;
using Poco::Net::SocketAddress;
using Poco::Net::SecureSMTPClientSession;
using Poco::Net::SMTPClientSession;
using Poco::SharedPtr;
using Poco::Net::MailMessage;
using Poco::Net::MailRecipient;
using namespace std;


int main(int argc, char** argv)
{
  string host = "smtp.gmail.com";
  int port = 465;
  std::string sUserName = "my_email_id";
  std::string sPassword = "my_password";
  std::string to = "recipient";
  std::string from = "my_email_id";
  std::string subject = "Your first e-mail message sent using Poco Libraries";
  std::string content = "Well done! You've successfully sent your first message using Poco SMTPClientSession";
  MailMessage message;

  try {
    SharedPtr<InvalidCertificateHandler> pCert = new AcceptCertificateHandler(false);
    Context::Ptr pContext = new Poco::Net::Context(Context::CLIENT_USE, "", "", "", Context::VERIFY_NONE, 9, false, "ALL:!ADH:!LOW:!EXP:!MD5:@STRENGTH");
    SSLManager::instance().initializeClient(0, pCert, pContext);

    SecureStreamSocket pSSLSocket(pContext);
    pSSLSocket.connect(SocketAddress(host, port));
    SecureSMTPClientSession secure(pSSLSocket);

    secure.login();
    bool tlsStarted = secure.startTLS(pContext);
    secure.login(SMTPClientSession::AUTH_LOGIN, sUserName, sPassword);

    message.setSender(from);
    message.addRecipient(MailRecipient(MailRecipient::PRIMARY_RECIPIENT, to));
    message.setSubject(subject);
    message.setContentType("text/plain; charset=UTF-8");
    message.setContent(content, MailMessage::ENCODING_8BIT);

     message.addAttachment("myFile", new FilePartSource("<file_path>"));

    secure.sendMessage(message);
    secure.close();

  } catch (Poco::Net::SMTPException &e) {
    cout << e.code() << endl;
    cout << e.message() << endl;
    cout << e.what() << endl;
    cout << e.displayText().c_str() << endl;
  }
  catch (Poco::Net::NetException &e) {
    cout << e.code() << endl;
    cout << e.message() << endl;
    cout << e.what() << endl;
    cout << e.displayText().c_str() << endl;
  }

  return 0;
}
```

The problem is that, when any file attachment is being sent, the content is not sent, i.e it does not show up in the receiver's email, or in the sent email in Gmail. However, when the attachment code is commented out and the mail is sent, the content shows up fine.

How do I do both?

**Comment by bakercp** [comment]
Are you sure that you aren't missing an exception from a missing file or something?

Also, if you are trying to send attachments, you shouldn't set the content as text/plain -- it should be a multi-part message.

Better to not set the content type at all and let the message take care if it automatically... here's an example.

https://github.com/bakercp/ofxSMTP/blob/master/example-advanced/src/ofApp.cpp#L109-L117

**Comment by Cupidvogel** [comment]
Thanks. I tried your code. I get three errors - two on the addContent and addAttachment lines, with the message - 

```C++
error: member reference type 'Poco::Net::MailMessage' is not a pointer; maybe you meant to use '.'?
        message->addAttachment(Poco::Net::MailMessage::encodeWord("<filePath","UTF-8")

error: member reference type 'Poco::Net::MailMessage' is not a pointer; maybe you meant to use '.'?
        message->addContent(new Poco::Net::StringPartSource("Hello world! How about an image?"))
```

And then, another one:

error: use of undeclared identifier 'ofToDataPath', but that I guess is because I don't have your included 'ofApp.h' file?

**Comment by bakercp** [comment]
Yeah, you'll need all of the requirements if you want to use my out of the box.  You can't just drop that code into yours.  It was intended to be an example of what methods are called to set up and send a multi-part message.

**Comment by Cupidvogel** [comment]
Okay, but what about the two not a pointer errors? Surely they are not due to your custom code?
 And as for the ofToDataPath method, what does it do?

**Comment by bakercp** [comment]
http://openframeworks.cc/documentation/utils/ofUtils.html#!show_ofToDataPath

https://stackoverflow.com/questions/4984600/when-do-i-use-a-dot-arrow-or-double-colon-to-refer-to-members-of-a-class-in-c

**Comment by Cupidvogel** [comment]
Oh damn. I did not observe that your message variable is not a pointer. I thought the problem was with the argument.. :)

**Comment by Cupidvogel** [comment]
We will always be using absolute paths to files. Why will the ofToDataPath method be still required then?

**Comment by bakercp** [comment]
It's not.  That is specific to my example (I didn't write the example for you, it is part of another package I maintain).  It is purely meant to show you the sequence of methods that I call to get the desired result.

**Comment by Cupidvogel** [comment]
Yep. I checked it, it works  fine with an absolute path as well. Thanks a lot! May be you can post your sample code in the official wiki (or the one you posted in my previous question, which I showed in this question), so that users can see how it works. Most of the sample codes on net regarding Poco does not work.. :(

**Comment by aleks-f** [comment]
> Most of the sample codes on net regarding Poco does not work.. :(

Most of the sample codes actually work fine for those who understand what they are doing. The discussion above shows clearly that there are things to learn before one can be productive in secure network C++ programming; that has nothing to do with quality of Poco examples. So, some reflection and self-examination (rather than biting the hand that's feeding you) may be appropriate.

**Comment by Cupidvogel** [comment]
Umm, I probably said thanks a million times, so I don't know what you mean by biting the hand that is feeding me. I saw the first Poco example here: http://axistasoft.com/blog/poco/poco-net/item/sending-email-messages-using-poco-smtpclientsession-class, and then here: http://rosettacode.org/wiki/Send_email#C.2B.2B, and given that neither of them uses secure connection at all, I don't see how they are valid examples and 'work fine for those who understand what they are doing'.  Yes, it is expected that someone planning to use a mailing library should have some basic knowledge of the internals (I have no problem in admitting that I am lacking there), but if they become a prerequisite to using the library, then may be better code examples could solve the problem, that's why I suggested to move a working pice of code to the wiki.

**Comment by bakercp** [comment]
I haven't had problems when I reference Poco's samples and tests (e.g.)

https://github.com/pocoproject/poco/blob/develop/Net/samples/Mail/src/Mail.cpp
https://github.com/pocoproject/poco/blob/develop/Net/samples/SMTPLogger/src/SMTPLogger.cpp

https://github.com/pocoproject/poco/blob/develop/Net/testsuite/src/MailMessageTest.cpp
https://github.com/pocoproject/poco/blob/develop/Net/testsuite/src/MailStreamTest.cpp

etc.

Your milage may vary when referencing 3rd party examples which often reference out-of-date versions of Poco - and the Poco team obviously can't fix 3rd party examples.

Is there Poco authored code you found that isn't working?

**Comment by Cupidvogel** [comment]
To be honest, I did not try these examples. But the first one refers to logging in without secure auth. I would have faced the same problem as I was initially facing when trying to authenticate via Gmail, right?

But yes, in the hindsight, I should have referred to the docs in built. Good advice to take. Thanks.

**Comment by aleks-f** [comment]
> To be honest, I did not try these examples. But the first one refers to logging in without secure auth.

And therein exactly lies the problem; not only you did not try examples, you disregarded my links pointing you to [secure email with Poco example](https://github.com/pocoproject/poco/blob/develop/NetSSL_OpenSSL/samples/Mail/src/Mail.cpp) over [a month ago](http://stackoverflow.com/questions/29302589/poco-unable-to-send-email/29358593#29358593).

When someone provides free code with documentation, examples etc. (with which perhaps you will even realize some monetary gain), and then you come asking questions already answered there, what you are implicitly saying is: "nice work you have there. now, let me take some more of your time for myself - instruct me, step by step, so I wouldn't have to put in the effort myself". The mere fact that Christopher had to send you links with explanation about use of dot/arrow/double-colon in C++, or that you would not even google for MX record but asked us to explain it to you proves it. Perhaps you were not aware of it or did not mean it intentionally, but that's how it comes across and that's what it effectively is. You do need to reconsider your approach to posting on technical forums; I give you this as a very honest feedback and a friendly advice.

**Comment by Cupidvogel** [comment]
Thanks. I will definitely consider it. To my defense, the code you pointed to over a month ago was already used by me and I even posted back the error I was getting - https://github.com/pocoproject/poco/issues/768, that first there was a connection refused error, then there were problems regarding port number, and finally the code suggested by Christopher worked. Before you suggested that code, I had already used another code (I cannot remember from where I got it, but not the Poco source for sure) which was very similar and it did not work as well.

I am new to both C++ and any networking related problem in general. So perhaps I was naive. But thanks for helping me out. I will do better. :)
